src/dune

```diff
@@ -15,6 +15,15 @@ Jbuild_plugin.V1.send
  (action (run ocaml %%{dep:../tools/gen_version.ml} Version.ml))
  (mode fallback))
 
+(rule
+ (copy#
+  compat/%%{read:compat-file}
+  compat.ml))
+
+(rule
+ (targets compat-file)
+ (action  (run %%{ocaml} %%{dep:../tools/gen_compat.ml} %%{ocaml_version})))
+
```

Comment:
I can't help but think that this mechanism looks a lot like dune's virtual libraries, where choosing one based on the compiler version could be a mild extension to the existing [Alternative dependencies](https://dune.readthedocs.io/en/latest/dune-files.html#alternative-deps) mechanism.